src/Tools/dotnet-format/CodeFormatter.cs

```diff
@@ -0,0 +1,131 @@
+                    await workspace.OpenProjectAsync(workspacePath, cancellationToken: token).ConfigureAwait(false);
+                    projectPath = workspacePath;
+                }
+
+                logger.LogTrace($"Workspace loaded in {stopwatch.ElapsedMilliseconds}ms.");
+                stopwatch.Restart();
+
+                await FormatFilesInSolutionAsync(logger, workspace.CurrentSolution, projectPath, token).ConfigureAwait(false);
+
+                logger.LogTrace($"Format complete in {stopwatch.ElapsedMilliseconds}ms.");
+            }
+
+            logger.LogInformation(Resources.Info_FormatComplete);
+
+            return 0;
+        }
+
+        private static async Task FormatFilesInSolutionAsync(ILogger logger, Solution solution, string projectPath, CancellationToken cancellationToken)
+        {
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                if (!string.IsNullOrEmpty(projectPath) && !project.FilePath.Equals(projectPath, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                logger.LogInformation(string.Format(Resources.Info_FormattingFilesInProject, project.Name));
+
+                if (project.Language != LanguageNames.CSharp && project.Language != LanguageNames.VisualBasic)
+                {
+                    logger.LogWarning(string.Format(Resources.Warn_UnsupportedProject, Path.GetFileName(project.FilePath)));
+                    continue;
+                }
+
+                var isCommentTrivia = project.Language == LanguageNames.CSharp
+                    ? (Func<SyntaxTrivia, bool>)IsCSharpCommentTrivia
+                    : IsVisualBasicCommentTrivia;
+
+                foreach (var documentId in project.DocumentIds)
+                {
+                    var document = project.GetDocument(documentId);
+                    if (!document.SupportsSyntaxTree)
+                    {
+                        continue;
+                    }
+
+                    var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                    if (GeneratedCodeUtilities.IsGeneratedCode(syntaxTree, isCommentTrivia, cancellationToken))
+                    {
+                        continue;
+                    }
+
+                    logger.LogTrace(string.Format(Resources.Info_FormattingFile, Path.GetFileName(document.FilePath)));
+
+                    var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                    var formattedDocument = await Formatter.FormatAsync(document, documentOptions, cancellationToken).ConfigureAwait(false);
+
+                    var documentText = await formattedDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    solution = solution.WithDocumentText(document.Id, documentText);
```

Comment:
this is super inefficient.  Each time we update a single documeent, we create the new solution snapshot.  If anything we do ends up needing semantics, this will cause a ton of work to be done over and over again.

Reply:
Indeed. Ran into this exact problem in the original codeformatter tool. Eventually had to stage the formats into the following:

1. Syntax only: can do all of these at one time without re-creating a `Solution`
2. Semantic which affected one project: allows you to only rebind the projects 
3. Semantic which affected whole solution. No choice but to rebind Solution after every change. 

---
In reply to: [211515699](https://github.com/dotnet/roslyn/pull/29403#discussion_r211515699) [](ancestors = 211515699)

Reply:
you can also change this to just be:  `solution = formattedDocument.Project.Solution;`  The new document already represents a full snapshot of hte world.

Reply:
@CyrusNajmabadi I tried the change you suggested and no changes were persisted to disk during the call to `.TryApplyChanges(solution)`. I am not sure what the difference is.

Reply:
can you debug through to find out?